File: 040_combination_sum_II.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int> > result;
        vector<int> path;
        sort(candidates.begin(),candidates.end());
        dfs(candidates,target,0,-1,path,result);
        return result;
    }
    void dfs(vector<int>& candidates, int target, int sum, int start, vector<int>& path, vector<vector<int> >& result) {
        if (sum>target) return;
        if (sum==target) {
            bool add=true;
            for (int index=0;index<result.size();index++) {
                if (path.size()==result[index].size() && path==result[index] ) {add=false; break;}
            }
            if (add) result.push_back(path);
        }
        for (int i=start+1;i<candidates.size();i++) {
            if (sum+candidates[i]>target) break;
            path.push_back(candidates[i]);
            dfs(candidates,target, sum+candidates[i],i,path,result);
            path.pop_back();
        }
        return;
        }
};

File: algorithms/python3/040_combination_sum_ii.py
class Solution:
    def combinationSum2(self, candidates, target):
        candidates.sort()
        res=[]
        def dfs(target,start,path):
            if target==0 and path not in res:
                res.append(path)
                return
            if target<0:
                return
            for i in range(start,len(candidates)):
                if candidates[i]>target:
                    break
                dfs(target-candidates[i],i+1,path+[candidates[i]])
        dfs(target,0,[])
        return res



a=Solution()

print(a.combinationSum2([10,1,2,7,6,1,5],8))
